### PR TITLE
Remove extra registry pod when it is created

### LINKED ISSUE #2613: e2e "config map update triggers registry pod rollout" failure
## Bug Report

e2e - "config map update triggers registry pod rollout" failed

Catalog represents a store of bundles which OLM can use to install Operators 
  config map update triggers registry pod rollout
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/catalog_e2e_test.go:203

[BeforeEach] Catalog represents a store of bundles which OLM can use to install Operators
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/catalog_e2e_test.go:40
[It] config map update triggers registry pod rollout
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/catalog_e2e_test.go:203
Creating catalog source mock-ocs-main-w9dzd in namespace operators...
Catalog source mock-ocs-main-w9dzd created
waiting for catalog pod mock-ocs-main-w9dzd to be available (for sync) - TRANSIENT_FAILURE
waiting for catalog pod mock-ocs-main-w9dzd to be available (for sync) - CONNECTING
waiting for catalog pod mock-ocs-main-w9dzd to be available (for sync) - TRANSIENT_FAILURE
waiting for catalog pod mock-ocs-main-w9dzd to be available (for sync) - TRANSIENT_FAILURE
waiting for catalog pod mock-ocs-main-w9dzd to be available (for sync) - CONNECTING
waiting for catalog pod mock-ocs-main-w9dzd to be available (for sync) - TRANSIENT_FAILURE
waiting for catalog pod mock-ocs-main-w9dzd to be available (for sync) - TRANSIENT_FAILURE
catalog mock-ocs-main-w9dzd pod with address mock-ocs-main-w9dzd.operators.svc:50051
catalog mock-ocs-main-w9dzd pod with address mock-ocs-main-w9dzd.operators.svc:50051
waiting for catalog pod mock-ocs-main-w9dzd to be available (after catalog update) - TRANSIENT_FAILURE
waiting for catalog pod mock-ocs-main-w9dzd to be available (after catalog update) - TRANSIENT_FAILURE
catalog updated
catalog updated
Waiting for pods matching selector olm.catalogSource=mock-ocs-main-w9dzd,olm.configMapResourceVersion=1769 to match given conditions
Waiting for pods matching selector olm.catalogSource=mock-ocs-main-w9dzd,olm.configMapResourceVersion=1769 to match given conditions
selector olm.catalogSource=mock-ocs-main-w9dzd,olm.configMapResourceVersion=1769 to match given conditions
[AfterEach] Catalog represents a store of bundles which OLM can use to install Operators
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/catalog_e2e_test.go:46
  cleaning up ephemeral test resources...
  deleting test subscriptions...
  deleting test installplans...
  deleting test catalogsources...
  deleting test crds...
  deleting test csvs...
  test resources delete

Failure [370.117 seconds]
Catalog represents a store of bundles which OLM can use to install Operators
/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/catalog_e2e_test.go:35
  config map update triggers registry pod rollout [It]
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/catalog_e2e_test.go:203
  	Error Trace:	util.go:108
  	            				catalog_e2e_test.go:293
  	            				runner.go:113
  	            				runner.go:64
  	            				it_node.go:26
  	            				spec.go:215
  	            				spec.go:138
  	            				spec_runner.go:200
  	            				spec_runner.go:170
  	            				spec_runner.go:66
  	            				suite.go:79
  	            				ginkgo_dsl.go:238
  	            				ginkgo_dsl.go:221
  	            				e2e_test.go:59
  	Error:      	Received unexpected error:
  	            	timed out waiting for the condition
  	Test:       	Catalog represents a store of bundles which OLM can use to install Operators config map update triggers registry pod rollout
  
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/vendor/github.com/stretchr/testify/require/require.go:1231

  Full Stack Trace
    github.com/stretchr/testify/require.NoError({0x7f51987718b0, 0xc001853cc0}, {0x3262c60, 0xc0001fe090}, {0x0, 0x0, 0x0})
  	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/vendor/github.com/stretchr/testify/require/require.go:1231 +0x96
  github.com/operator-framework/operator-lifecycle-manager/test/e2e.awaitPods({0x330c978, 0xc001853cc0}, {0x33272e0, 0xc001012090}, {0x7ffd1a2eee5e, 0x9}, {0xc000f15040, 0x47}, 0xc000fdacd0)
    	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/util.go:108 +0x215
  github.com/operator-framework/operator-lifecycle-manager/test/e2e.glob..func3.5()
  	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/catalog_e2e_test.go:293 +0x15ae
  github.com/operator-framework/operator-lifecycle-manager/test/e2e.TestEndToEnd(0xf908f9)
  	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/e2e_test.go:59 +0x268
  testing.tRunner(0xc000ab04e0, 0x3004318)
  	/opt/hostedtoolcache/go/1.17.6/x64/src/testing/testing.go:1259 +0x102
  created by testing.(*T).Run
  	/opt/hostedtoolcache/go/1.17.6/x64/src/testing/testing.go:1306 +0x35a

**What did you do?**
Check CI e2e failure
**What did you expect to see?**
No errors
**What did you see instead? Under which circumstances?**
Failure above

**Environment**
* operator-lifecycle-manager version:

* Kubernetes version information:

* Kubernetes cluster kind: 

**Possible Solution**

**Additional context**

### PR DESCRIPTION
**Pull request by akihikokuroda**
Signed-off-by: akihikokuroda <akihikokuroda2020@gmail.com>

**Description of the change:**

This PR adds a mutex block around the creation of the configmap retistry pod.  I reproduced this error locally a couple of times.  I saw 2 registry pods created with the same **configMapResourceVersion**.  The 2 queue process workers created the pod at the same time.  This mutex block prevent 2 workers checking and creating the registry pod at the same time. 

**Motivation for the change:**
Closes #2613 
**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

**Comment by akihikokuroda** [comment]
/hold
The mutex block doesn't seem enough because of the cached pod instances.

**Comment by akihikokuroda** [comment]
/unhold
Change the currentPod function not to use cache to get the list of pods.

File: pkg/controller/registry/reconciler/configmap.go
Comment (njhale) [inline pkg/controller/registry/reconciler/configmap.go:170]: I don't think this mutex needs to be exported.

File: pkg/controller/registry/reconciler/configmap.go
Comment (njhale) [inline pkg/controller/registry/reconciler/configmap.go:364]: It's not clear to me that adding a mutex solves the problem at hand. The queues that are used key on name/namespace and should never allow the same resource to be processed by more than one worker concurrently.

IMO, an alternative suggestion for the behavior you're seeing is that the resource was processed more than once, serially, before the cache that backs the `Lister` -- which you've swapped with a client call above --  can be updated. This could have lead to cache misses, and the resource being recreated.

I would feel much more comfortable if you could produce a test, unit or e2e, that fails for the original code but passes with these changes.

**Review by njhale — CHANGES_REQUESTED** [review]
Thanks for submitting this. I have a few questions/comments before we merge:

File: pkg/controller/registry/reconciler/configmap.go
Comment (akihikokuroda) [reply pkg/controller/registry/reconciler/configmap.go:364]: Thanks for review.  I saw 2 pods were created so I guessed the cause of it.  I'll look into more detail and capture what is happening.  I'm not sure if I can create a test to reproduce consistently but I'll try it anyway.

**Comment by akihikokuroda** [comment]
It must be the cache misses.  Here is the catalog operator pod log when the CI test failed.  The  **catsrc id=wEdnr** and **catsrc id=iFe3m** ran back to back and the **catsrc id=iFe3m** saw the `check registry server healthy: false` after the `ensured registry server` by **catsrc id=wEdnr**

```
time="2022-01-28T18:39:08Z" level=debug msg="syncing catsrc" id=wEdnr source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=debug msg="checking catsrc configmap state" id=wEdnr source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=debug msg="check registry server healthy: false" id=wEdnr source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=debug msg="ensuring registry server" id=wEdnr source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=info msg="syncing catalog source for annotation templates" catSrcName=mock-ocs-main-w9dzd catSrcNamespace=operators id=ev0O2
time="2022-01-28T18:39:08Z" level=debug msg="this catalog source is not participating in template replacement" catSrcName=mock-ocs-main-w9dzd catSrcNamespace=operators id=ev0O2
time="2022-01-28T18:39:08Z" level=debug msg="RemoveStatusConditions - request to remove status conditions did not result in any changes, so updates were not made" catSrcName=mock-ocs-main-w9dzd catSrcNamespace=operators id=ev0O2
time="2022-01-28T18:39:08Z" level=debug msg="Got source event: grpc.SourceState{Key:registry.CatalogKey{Name:\"mock-ocs-main-w9dzd\", Namespace:\"operators\"}, State:1}"
time="2022-01-28T18:39:08Z" level=info msg="state.Key.Namespace=operators state.Key.Name=mock-ocs-main-w9dzd state.State=CONNECTING"
time="2022-01-28T18:39:08Z" level=debug msg="Got source event: grpc.SourceState{Key:registry.CatalogKey{Name:\"mock-ocs-main-w9dzd\", Namespace:\"operators\"}, State:3}"
time="2022-01-28T18:39:08Z" level=info msg="state.Key.Namespace=operators state.Key.Name=mock-ocs-main-w9dzd state.State=TRANSIENT_FAILURE"
time="2022-01-28T18:39:08Z" level=debug msg="ensured registry server" id=wEdnr source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=debug msg="syncing catsrc" id=iFe3m source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=debug msg="checking catsrc configmap state" id=iFe3m source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=debug msg="check registry server healthy: false" id=iFe3m source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=debug msg="ensuring registry server" id=iFe3m source=mock-ocs-main-w9dzd
time="2022-01-28T18:39:08Z" level=info msg="syncing catalog source for annotation templates" catSrcName=mock-ocs-main-w9dzd catSrcNamespace=operators id=sY5K/
time="2022-01-28T18:39:08Z" level=debug msg="this catalog source is not participating in template replacement" catSrcName=mock-ocs-main-w9dzd catSrcNamespace=operators id=sY5K/
time="2022-01-28T18:39:08Z" level=debug msg="RemoveStatusConditions - request to remove status conditions did not result in any changes, so updates were not made" catSrcName=mock-ocs-main-w9dzd catSrcNamespace=operators id=sY5K/
time="2022-01-28T18:39:08Z" level=debug msg="ensured registry server" id=iFe3m source=mock-ocs-main-w9dzd
```
I made change to not to use the Lister to check the registry pod in the **currentPods** and **currentPodsWithCorrectResourceVersion**.  This should make the second sync to see the registry pod created.
I couldn't make the tests that causes this failure with the original code.  It seems a small timing gap case.

File: pkg/controller/registry/reconciler/configmap.go
Comment (akihikokuroda) [reply pkg/controller/registry/reconciler/configmap.go]: It is taken out.

File: pkg/controller/registry/reconciler/configmap.go
Comment (ecordell) [inline pkg/controller/registry/reconciler/configmap.go:217]: This may technically solve the issue, but if at all possible, talking directly to the kube api for `read` requests should be avoided.

The operator is already watching pods, so it will eventually know about all of these pods, and this just adds load to the apiserver.

Comment (akihikokuroda) [reply pkg/controller/registry/reconciler/configmap.go:217]: Thanks for comments.  OK.  The second pod stays up unused until the next change of the catalogsource.  I can probably change code to stop the second pod sooner without issuing the read requests instead preventing the second pod creation.

**Comment by perdasilva** [comment]
closing PR as stale. Please re-open if it's still important.

**Pull request closed without merging**
